### PR TITLE
add tx_priority_fee argument for program deploy subcommand

### PR DESCRIPTION
**Pull request by Nagaprasadvr**
Issue - #21 

Add Transaction priority fee argument in terms of LAMPORTS to solana program deploy subcommand

#### Problem
1. solana program deploy command could take a lot of time to finish due to program code size .Transaction execution speeds can be increased or txs can be prioritised by setting compute_unit_price (priority fee) for every tx this price is set and txs are executes faster than normal therefore making program deployment faster. 
usage - solana program deploy --tx-priority-fee 1000000 

#### Summary of Changes

solana/cli/program - added tx_priority_fee logic

### CODE REVIEW

**Comment by Nagaprasadvr** [comment]
Should we update program_v4 as well ?

File: cli/src/program.rs [hunk 1]
```diff
@@ -2364,13 +2429,23 @@ fn do_process_program_upgrade(
...
 ) -> ProcessResult {
     let blockhash = rpc_client.get_latest_blockhash()?;
 
     let (initial_message, write_messages, balance_needed) =
         if let Some(buffer_signer) = buffer_signer {
+            let mut initial_instructions: Vec<Instruction> = Vec::new();
+
+            // Set compute unit price
+            if let Some(tx_priority_fee) = tx_priority_fee {
+                initial_instructions.push(ComputeBudgetInstruction::set_compute_unit_price(
```
Comment (tao-stones) [inline cli/src/program.rs]: nit: perhaps rename the new arg to `compute-unit-price` (instead of `tx_priority_fee`) to be consistent with `CompuyeBudgetInstruction` it invokes.

Comment (Nagaprasadvr) [reply cli/src/program.rs:2443]: yup will do

Comment (Nagaprasadvr) [reply cli/src/program.rs]: resolved

File: cli/src/program.rs [hunk 2]
```diff
@@ -2253,6 +2315,8 @@ fn do_process_program_write_and_deploy(
             min_rent_exempt_program_data_balance,
         )
     };
+
+    initial_instructions.extend(ixs);
```
Comment (tao-stones) [inline cli/src/program.rs:2282]: It could error out when making `ixs` above, so maybe more economic to extend `ComputeBudgetInstruction::set_compute_unit_price(...)` into `ixs`, instead of allocate set_compute_unit_price then fails on rpc call.

File: cli/src/program.rs [hunk 3]
```diff
@@ -2209,11 +2261,21 @@ fn do_process_program_write_and_deploy(
...
     skip_fee_check: bool,
+    max_retries: Option<usize>,
+    compute_unit_price: Option<u64>,
 ) -> ProcessResult {
     let blockhash = rpc_client.get_latest_blockhash()?;
+    let mut initial_instructions: Vec<Instruction> = Vec::new();
+
+    // Set compute unit price
+    if let Some(compute_unit_price) = compute_unit_price {
+        initial_instructions.push(ComputeBudgetInstruction::set_compute_unit_price(
```
Comment (tao-stones) [inline cli/src/program.rs]: it is highly recommended to set `compute-unit-price` and `compute-unit-limit` together. Using default cu-limit may ended up to be overpaying

File: cli/src/program.rs [hunk 4]
```diff
@@ -236,6 +242,23 @@ impl ProgramSubCommands for App<'_, '_> {
...
+                                    [Note: priority fee is added in LAMPORTS, so enter priority fee in terms of LAMPORTS]",
+                                ),
+                        ).
+                        arg(
+                            Arg::with_name("max_retires")
+                                .long("max-retries")
+                                .takes_value(true)
+                                .help(
+                                    "Specifies the max retry count for failed transactions",
+                                ),
```
Comment (t-nelson) [inline cli/src/program.rs]: these should be added in separate PRs

File: cli/src/program.rs [hunk 5]
```diff
@@ -236,6 +242,23 @@ impl ProgramSubCommands for App<'_, '_> {
...
+                        )
+                        .arg(
+                            Arg::with_name("compute_unit_price")
+                                .long("compute-unit-price")
+                                .takes_value(true)
+                                .help(
+                                    "Set computeUnitPrice (tx priotity fee) to boost transactions resulting in faster execution speed \
+                                    [Note: priority fee is added in LAMPORTS, so enter priority fee in terms of LAMPORTS]",
+                                ),
+                        ).
```
Comment (t-nelson) [inline cli/src/program.rs]: we'll want to define this arg somewhere and such that it can be reused on all transaction-bearing subcommands, for consistencies sake. most likely that will be in the `clap-utils` dir, similarly to the `offline_signing` module

File: cli/src/program.rs (hunk 4 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: so compute_unit_price is one pr and max_retries is another pr ?

File: cli/src/program.rs (hunk 5 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: can we do this in a diff pr ?

File: cli/src/program.rs (hunk 4 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: resolved

File: cli/src/program.rs (hunk 2 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs:2282]: After reading this https://www.helius.dev/blog/priority-fees-understanding-solanas-transaction-fee-mechanics Section:Best practices , it says set_compute_limit and set_compute_units should  always be added to the start of ix Vec.

So a solution for this can be i create set_compute_units and set_compute_limit instructions after `ixs` is prepared ? 
at line no 2318

File: cli/src/program.rs (hunk 3 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: this is resolved , i calculate compute_units by multiplying ixs.len() + 2 with 200k 
then set compute_limit

**Comment by ChewingGlass** [comment]
FYI this seems to only put priority fees on the initial buffer creation transaction and not on all the subsequent txns (the ones that take the lion's share of the time)

**Comment by Nagaprasadvr** [comment]
> FYI this seems to only put priority fees on the initial buffer creation transaction and not on all the subsequent txns (the ones that take the lion's share of the time)

I have added set compute limit and set compute price ixs to write messages and final messages as well

File: cli/src/program.rs [hunk 6]
```diff
@@ -2275,7 +2319,16 @@ fn do_process_program_write_and_deploy(
         } else {
             loader_instruction::write(buffer_pubkey, loader_id, offset, bytes)
         };
-        Message::new_with_blockhash(&[instruction], Some(&fee_payer_signer.pubkey()), &blockhash)
+        if let Some(compute_unit_price) = compute_unit_price {
+            ixs.push(ComputeBudgetInstruction::set_compute_unit_limit(
+                3 as u32 * 200_000u32,
+            ));
+            ixs.push(ComputeBudgetInstruction::set_compute_unit_price(
+                0u64.mul((10 as u64).pow(6)),
```
Comment (ChewingGlass) [inline cli/src/program.rs]: Will always be 0?

Comment (ChewingGlass) [reply cli/src/program.rs]:
```suggestion
                compute_unit_price.mul((10 as u64).pow(6)),
```

**Comment by ChewingGlass** [comment]
The arg should be in terms of microlamports, otherwise the min you can pay is 0.000605 SOL per write tx. Which gets pricy.

File: cli/src/program.rs (hunk 6 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: > Will always be 0?

Oh my bad , that is wrong it should be compute_unit_price

**Comment by Nagaprasadvr** [comment]
> The arg should be in terms of microlamports, otherwise the min you can pay is 0.000605 SOL per write tx. Which gets pricy.

**Comment by Nagaprasadvr** [comment]
> The arg should be in terms of microlamports, otherwise the min you can pay is 0.000605 SOL per write tx. Which gets pricy.

Will change to micro lamports

File: cli/src/program.rs (hunk 6 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: resolved

**Comment by Nagaprasadvr** [comment]
> The arg should be in terms of microlamports, otherwise the min you can pay is 0.000605 SOL per write tx. Which gets pricy.

resolved

File: cli/src/program.rs [hunk 7]
```diff
@@ -236,7 +239,16 @@ impl ProgramSubCommands for App<'_, '_> {
                                     "Use the designated program id even if the account already \
                                      holds a large balance of SOL",
                                 ),
-                        ),
+                        )
+                        .arg(
+                            Arg::with_name("compute_unit_price")
```
Comment (t-nelson) [inline cli/src/program.rs]: prefer the pre-defined `Arg` from `clap-utils`

https://github.com/anza-xyz/agave/blob/a2579d484efd78099ba67db5911b2ba12c77b711/clap-utils/src/compute_unit_price.rs#L9-L15

File: cli/src/program.rs [hunk 8]
```diff
@@ -2730,6 +2784,22 @@ fn report_ephemeral_mnemonic(words: usize, mnemonic: bip39::Mnemonic) {
     eprintln!("[BUFFER_ACCOUNT_ADDRESS] argument to `solana program close`.\n{divider}");
 }
 
+fn set_compute_budget_ixs_if_needed(
+    ixs: &mut Vec<Instruction>,
+    compute_unit_price: &Option<u64>,
+    ixs_len: usize,
+) {
+    if let Some(compute_unit_price) = compute_unit_price {
+        println!("ixs_len: {}", ixs_len);
```
Comment (t-nelson) [inline cli/src/program.rs]: remove this debug print

File: cli/src/program.rs [hunk 9]
```diff
@@ -2730,6 +2784,22 @@ fn report_ephemeral_mnemonic(words: usize, mnemonic: bip39::Mnemonic) {
...
+        ixs.push(ComputeBudgetInstruction::set_compute_unit_limit(
+            ((ixs_len + 2) as u32) * 200_000u32,
```
Comment (t-nelson) [inline cli/src/program.rs]: most likely this is a _drastic_ over-reservation which may reduce the likelihood of the transaction being included toward the end of a block

File: cli/src/program.rs (hunk 8 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: resolved

File: cli/src/program.rs (hunk 7 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: resolved

File: cli/src/program.rs (hunk 9 above)
Comment (Nagaprasadvr) [reply cli/src/program.rs]: Can u elaborate what is wrong here exactly is it setting more cu than required?

i have also read that setting compute_unit_limit is optional when setting compute_unit_price wyt?

Comment (jstarry) [reply cli/src/program.rs]: It is optional, but it's recommended to set a limit because the default limit is 200k * the number of instructions in the transaction. The transaction priority fee is calculated by multiplying the micro-lamport fee (per compute unit) by the transaction's compute unit limit. So if you set the compute unit too high or if you don't specify one and use the default limit, you will end up overpaying priority fees.. in other words, wasting tokens.

This `set_compute_budget_ixs_if_needed` should probably accept `compute_unit_limit` argument explicitly and then each caller can calculate the exact compute unit limit needed for the passed `ixs`. Then this function can add 300 compute units to the limit for the two additional compute budget program instructions. The default compute units charged per core program instruction are as follows:
- System program: 150
- Compute budget program: 150
- Upgradeable BPF loader program: 2370

Comment (Nagaprasadvr) [reply cli/src/program.rs]: will add this logic ty

Comment (Nagaprasadvr) [reply cli/src/program.rs]: so we need to check each ix in ix vec and calculate compute_units required in this case right?

Comment (jstarry) [reply cli/src/program.rs]: Yes each ix has to be counted. I already explained this above and gave my suggestion for how to implement the fix.

**Comment by Nagaprasadvr** [comment]
found a workaround

**Comment by Nagaprasadvr** [comment]
found a workaround for this
@jstarry 

i have added 2000 cu when program is neither system_program or bpf_upgradable_loader
image

**Comment by Nagaprasadvr** [comment]
please look at #312 ! apologies for closing  this pr , please add your comments there
